### PR TITLE
Consolidate diagnostics IDs for "use builtin/framework types".

### LINKED ISSUE #27042: Consolidate diagnostic IDs for predefined types
**Version Used**: The proposal is an implementation detail of #26578 which is called to the surface.

:memo: This change has since been removed from #26578, but is still under consideration as an independent change.

**Current behavior:**

Currently diagnostics related to predefined types have four diagnostic IDs. Suppose a user wishes to use `string` when declaring a variable with this type, but `String` when referring to static members. The following code would produce the two diagnostics:

```csharp
String value // IDE0012
  = string.Empty; // IDE0015
```

Based on the configuration selected by the user, only two distinct IDs can appear in any given file (.editorconfig allows different configurations to apply to different source files, so it's possible a project or solution could reveal all four).

**Proposed behavior:**

The proposed behavior would consolidate all four situations to use the same diagnostic ID. This diagnostic ID would be a different ID from the previous ones, and the current IDs would be retired. The configuration functionality would not change *at all*. The benefit of this change is it allows Fix All operations to fix *all* situations related to this configuration in a single pass, rather than requiring users to make up to four passes to fix everything.

**Breaking change:**

This is a breaking change for cases where diagnostics are suppressed. However, this is a relatively low risk change because the impacted analyzers were never shipped as NuGet packages and thus have never had an impact on the actual build outcome.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/27042.

This is a followup to https://github.com/dotnet/roslyn/pull/26578 and https://github.com/dotnet/roslyn/pull/26825.  I personally think those should go in first.  However, @sharwell made a reasonable case that we might want to just deliver this all in one go.

### Customer scenario

Completes the approved design from #27042. If we do not take this change, 15.8 Preview 4 will change the meaning of some diagnostic IDs related to this feature to a regression state which was never intended to ship.

### Bugs this fixes

#27042

### Workarounds, if any

None.

### Risk

N/A. This was part of the design work in #26578 and cannot be separated from it.

### Performance impact

N/A. This was part of the design work in #26578 and cannot be separated from it.

### Is this a regression from a previous update?

Failure to include this change will introduce a regression.

### Root cause analysis

Infrastructure stability problems leading to higher approval mode delayed a pull request which was expected to merge several days earlier.

### How was the bug found?

N/A (known tracked issue)

### Test documentation updated?

No.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @sharwell

**Comment by CyrusNajmabadi** [comment]
Also tagging @jcouv @mavasani @dotnet/roslyn-ide @heejaechang

**Comment by CyrusNajmabadi** [comment]
@sharwell How would you like to proceed on these PRs?  Doing them one at a time?  Or just take this larger single PR?

File: src/EditorFeatures/VisualBasicTest/Diagnostics/PreferFrameworkType/PreferFrameworkTypeTests_FixAllTests.vb [hunk 1]
```diff
@@ -348,9 +346,9 @@ End Class]]>
                                    <Document><![CDATA[
 Imports System
 Class ProgramA
-    Dim x As Integer = 0
-    Dim y As Integer = 0
-    Dim z As Integer = 0
+    Dim x As Int32 = 0
+    Dim y As Int32 = 0
+    Dim z As Int32 = 0
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/VisualBasicTest/Diagnostics/PreferFrameworkType/PreferFrameworkTypeTests_FixAllTests.vb]: This behavior is what we want.  We've consolidated to a single ID, and so now we update all of these when someone does 'fix all'.

**Comment by CyrusNajmabadi** [comment]
@sharwell This is ready for review.

File: src/EditorFeatures/CSharpTest/Diagnostics/PreferFrameworkType/PreferFrameworkTypeTests_FixAllTests.cs [hunk 2]
```diff
@@ -412,9 +412,9 @@ private System.Int32 F(System.Int32 p1, System.Int16 p2)
 using System;
 class ProgramA
 {
-    private int x = 0;
-    private int y = 0;
-    private int z = 0;
+    private Int32 x = 0;
+    private Int32 y = 0;
+    private Int32 z = 0;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/PreferFrameworkType/PreferFrameworkTypeTests_FixAllTests.cs]: This behavior is what we want. We've consolidated to a single ID, and so now we update all of these when someone does 'fix all'.

File: src/EditorFeatures/CSharpTest/SimplifyTypeNames/SimplifyTypeNamesTests.cs [hunk 3]
```diff
@@ -1201,6 +1201,20 @@ static void Main(string[] args)
...
+class Program
+{
+    static void Main(string[] args)
+    {
+        var x = nameof(typeof(int));
+    }
+
     static string nameof(Type t)
     {
         return string.Empty;
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/SimplifyTypeNames/SimplifyTypeNamesTests.cs]: this test seems to ahve been bogus.  It was trying to validate that we were not converting 'Int32' to 'int' inside of a nameof.  However, in this particular test, we don't have a real 'nameof'.  Instead, we have a method called 'nameof'.  In that case we *do* want to simplify to 'int' as it is legal here.

File: src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpSimplifyTypeNamesDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -93,13 +91,23 @@ private static bool IsCrefCandidate(SyntaxNode node)
...
+                model, node, optionSet,
+                out issueSpan, out diagnosticId, out inDeclaration,
+                cancellationToken);
         }
 
-        internal override bool CanSimplifyTypeNameExpression(SemanticModel model, SyntaxNode node, OptionSet optionSet, out TextSpan issueSpan, out string diagnosticId, CancellationToken cancellationToken)
+        internal override bool CanSimplifyTypeNameExpression(
+            SemanticModel model, SyntaxNode node, OptionSet optionSet,
+            out TextSpan issueSpan, out string diagnosticId, out bool inDeclaration,
+            CancellationToken cancellationToken)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpSimplifyTypeNamesDiagnosticAnalyzer.cs]: only added ```out bool inDeclaration```

**Comment by CyrusNajmabadi** [comment]
@sharwell Can i get a review on this?

File: src/EditorFeatures/CSharpTest/SimplifyTypeNames/SimplifyTypeNamesTests.cs [hunk 5]
```diff
@@ -1191,7 +1191,7 @@ static void Main(string[] args)
         [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsSimplifyTypeNames)]
         public async Task SimplifyToPredefinedTypeNameShouldNotBeOfferedInsideNameOf4()
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/SimplifyTypeNames/SimplifyTypeNamesTests.cs]: Update the test name too?

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/SimplifyTypeNames/SimplifyTypeNamesTests.cs]: That would make too much sense :)

File: src/Features/Core/Portable/Diagnostics/Analyzers/IDEDiagnosticIds.cs [hunk 6]
```diff
@@ -72,6 +75,9 @@ internal static class IDEDiagnosticIds
         public const string RemoveUnnecessaryParenthesesDiagnosticId = "IDE0047";
         public const string AddRequiredParenthesesDiagnosticId = "IDE0048";
 
+        public const string PreferBuiltInOrFrameworkTypeDiagnosticId = "IDE0049";
```
Comment (jcouv) [inline src/Features/Core/Portable/Diagnostics/Analyzers/IDEDiagnosticIds.cs]: Nit: extra empty line below

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Diagnostics/Analyzers/IDEDiagnosticIds.cs]: I can clean in a subsequent PR :)  (i have a bunch that touch this file) i would prefer to not have to reset this one :D

**Comment by CyrusNajmabadi** [comment]
@jcouv Any other concerns?

**Comment by CyrusNajmabadi** [comment]
Thanks guys .  @sharwell @jinujoseph to merge when you want.

**Comment by sharwell** [comment]
retest windows_debug_spanish_unit32_prtest please

**Comment by sharwell** [comment]
retest this please

**Comment by CyrusNajmabadi** [comment]
Thanks guys . @sharwell @jinujoseph to merge when you want.

**Comment by jinujoseph** [comment]
waiting on test

**Comment by CyrusNajmabadi** [comment]
@jinujoseph 
> waiting on test

How long do tests take?  The last commit was 4 days ago :)

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by CyrusNajmabadi** [comment]
@jinujoseph All tests have passed.  merge when convenient.

**Comment by CyrusNajmabadi** [comment]
@jinujoe Any movement on this?

**Comment by CyrusNajmabadi** [comment]
er: @jinujoseph

**Comment by jinujoseph** [comment]
Approved to merge for 15.8.Preview4

**Comment by CyrusNajmabadi** [comment]
Thanks!

**Comment by jinujoseph** [comment]
Thanks @CyrusNajmabadi  for fixing this!

**Comment by sharwell** [comment]
🎉

**Pull request merged**